lib/Sema/CSSolver.cpp

```diff
@@ -1020,11 +1020,30 @@ void ConstraintSystem::shrink(Expr *expr) {
   }
 }
 
+static bool debugConstraintSolverForExpr(ASTContext &C, Expr *expr) {
+  if (C.LangOpts.DebugConstraintSolver)
+    return true;
+
+  if (C.LangOpts.DebugConstraintSolverOnLines.empty())
+    // No need to compute the line number to find out it's not present.
+    return false;
+
+  unsigned line = 0;
+  if (expr->getLoc().isValid())
+    line = C.SourceMgr.getLineNumber(expr->getLoc());
```

Comment:
I don't think this works when there is a single statement closure and line number points to its body e.g.

```foo(..) {
  _ = x + y <- this line
}
```

Since closure is type-checked together with enclosing context there is going to be no debug output. Also consider expressions with implicit containers (SanitizeExpr would mutate AST)? I think this method needs to evaluate whole AST to figure out if debug should be enabled or not.

Reply:
My goal was to use the same line number `solveImpl()` logs when it starts solving an expression. I guess that should actually be `getStartLoc()`, though.

Reply:
I've modified it to calculate both the start and end lines of the expression and activate debugging if any line in the list is between the two.

Reply:
Perfect, thank you!